sourcecode/apis/contentauthor/tests/Integration/Libraries/H5P/AjaxRequestTest.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace Tests\Integration\Libraries\H5P;
+
+use App\H5PLibrary;
+use App\H5PLibraryLibrary;
+use App\Libraries\ContentAuthorStorage;
+use App\Libraries\H5P\AjaxRequest;
+use App\Libraries\H5P\Framework;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+use Tests\TestCase;
+
+class AjaxRequestTest extends TestCase
+{
+    use RefreshDatabase;
+
+    public function test_libraryRebuild(): void
+    {
+        /** @var H5PLibrary $library */
+        $library = H5PLibrary::factory()->create();
+        /** @var H5PLibrary $preLib */
+        $preLib = H5PLibrary::factory()->create(['name' => 'player', 'major_version' => 3, 'minor_version' => 14]);
+        /** @var H5PLibrary $dynLib */
+        $dynLib = H5PLibrary::factory()->create(['name' => 'H5P.Dynamic', 'major_version' => 2, 'minor_version' => 42, 'patch_version' => 3, 'patch_version_in_folder_name' => true]);
+        /** @var H5PLibrary $edLib */
+        $edLib = H5PLibrary::factory()->create(['name' => 'FontOk', 'major_version' => 1, 'minor_version' => 3]);
+
+        H5PLibraryLibrary::create(['library_id' => $library->id, 'required_library_id' => $preLib->id, 'dependency_type' => 'preloaded']);
+        H5PLibraryLibrary::create(['library_id' => $library->id, 'required_library_id' => $dynLib->id, 'dependency_type' => 'dynamic']);
+        H5PLibraryLibrary::create(['library_id' => $library->id, 'required_library_id' => $edLib->id, 'dependency_type' => 'editor']);
+
+        $storage = $this->createMock(ContentAuthorStorage::class);
+        $this->instance(ContentAuthorStorage::class, $storage);
+        $storage
+            ->expects($this->exactly(4))
+            ->method('copyFolder');
+
+        $framework = $this->createMock(Framework::class);
```

Comment:
Why are mocks needed here?

I'm concerned this test will break when one of the assumptions as to how a library is rebuilt changes, without this necessarily meaning that rebuilding itself is broken. The mock approach makes the test hard to understand and keep updated, so it would either take ages to fix, or we'd do `markTestIncomplete()` and call it a day.

The preferred approach would be to check that rebuilding had the expected effect without checking how that effect was achieved, e.g. by using Laravel's facilities for database and filesystem assertions prior to and after rebuilding. But perhaps this isn't possible for some reaosn?

Reply:
Most of the mocks are removed